### ISSUE TITLE
LightGbmRanker: Significance of having TrainingData rows sorted with respect to Rank order.

### ISSUE THREAD

**Issue by Sumanth9Nandamuri** [dotnet/machinelearning#6157]
### System information

.NET SDK (reflecting any global.json):
 Version:   6.0.202
 Commit:    f8a55617d2

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19044
 OS Platform: Windows
 RID:         win10-x64

### Issue

- **What did you do?**
Train and evaluate LightGbmRanker using ML.Net C# API.
Training and test data are clustered/grouped with respect to GroupId column.

- **What happened?**
**Experiment 1:** Rows in Training data are sorted with respect to RankLabel, in this case rows corresponding to a GroupId have RankLabel 13, 12, 11, ........1, 1, 1, ..etc (Max rank is 13 for our use case)
**Experiment 2:** Rows in Training data are not sorted with respect to RankLabel, in this case rows corresponding to a GroupId have RankLabel in arbitrary order 
**TestData:** Even though test data has a RankLabel (for evaluation/ metric computation purpose), the rows corresponding to a GroupId are not sorted in any order.
- **Results:**
In case of experiment 1, NDCG score on TestData is quite high (around 0.95)
In case of experiment 2, NDCG score on TestData is quite low (around 0.25)

- **What did you expect?**
Same NDCG Score from both experiments.

### Source code / logs

using Microsoft.ML;
using Microsoft.ML.Data;
using Microsoft.ML.Transforms;
using Microsoft.ML.Trainers.LightGbm;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Diagnostics;

namespace Ranking
{
    class Program
    {
        static void Main(string[] args)
        {
            int maxRank = 13;

            var loaderOptions = new TextLoader.Options
            {
                AllowQuoting = true,
                HasHeader = false,
                Separators = new char[] {'\t'}
            };

            double lr = 0.05;
            int n_leaves = 30;
            int min_example_count_per_leaf = 50;
            string param_string = $"Dart_{n_leaves}_{min_example_count_per_leaf}_{lr}";
            var boosting = new DartBooster.Options();
            boosting.XgboostDartMode = true;
            
            var trainerOptions = new LightGbmRankingTrainer.Options
            {
                RowGroupColumnName = "GroupId",
                FeatureColumnName = "Features",
                LabelColumnName = "Label",
                BatchSize = 1048576,
                CustomGains = new int[] {1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096},
                HandleMissingValue = true,
                EvaluationMetric = LightGbmRankingTrainer.Options.EvaluateMetricType.NormalizedDiscountedCumulativeGain,
                LearningRate = lr,
                MaximumBinCountPerFeature = 255,
                MinimumExampleCountPerLeaf = min_example_count_per_leaf,
                NumberOfLeaves = n_leaves,
                Seed = 42,
                Booster = boosting,
                EarlyStoppingRound = 0,
                NumberOfIterations = 100
            };

            var evaluatorOptions = new RankingEvaluatorOptions
            {
                DcgTruncationLevel = maxRank,
                LabelGains = "1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096"
            };

            StreamWriter sw = new StreamWriter($"../Logs/LightGbm_{param_string}.txt");
            sw.WriteLine("=> Training Model");
            sw.WriteLine(" ");
            sw.WriteLine(" ");
            
            sw.WriteLine($"=> Printing Logs}");

            var watch = Stopwatch.StartNew();
            ReadOnlyList<double> ndcg = Train(loaderOptions, trainerOptions, evaluatorOptions);
            watch.Stop();
            var elapsedMs = watch.ElapsedMilliseconds;
            sw.WriteLine($"=> Model trained and saved");
            sw.WriteLine($"=> Metrics");
            for(int i=0; i<maxRank; i++)
            {
                sw.WriteLine($"=> NDCG@{i+1} :  {ndcg[i]}");
            }
            sw.WriteLine($"=> Averaged NDCG: {ndcg.Average()}");
            sw.WriteLine($"=> Time Elapsed: {elapsedMs/1000} sec");

            Console.WriteLine($"=> Done. || Averaged NDCG: {ndcg.Average()} || Time elapsed {elapsedMs/1000} sec");
            }

            sw.Flush();
            sw.Close();
        }

        static IReadOnlyList<double> Train(
                          TextLoader.Options loaderOptions, 
                          LightGbmRankingTrainer.Options trainerOptions,
                          RankingEvaluatorOptions evaluatorOptions)
        {
            MLContext mlContext = new MLContext(seed: 42);

            IDataView trainingData = mlContext.Data.LoadFromTextFile<DataSchema>($"../RankingData/Train/RankingData_Train_RandOrder.tsv", loaderOptions);
            IDataView testData = mlContext.Data.LoadFromTextFile<DataSchema>($"../RankingData/Test/RankingData_Test.tsv", loaderOptions);

            var pipeline = mlContext.Transforms.DropColumns(new string[] {"C1"})
                           .Append(mlContext.Transforms.Categorical.OneHotHashEncoding("GroupId", outputKind: OneHotEncodingEstimator.OutputKind.Key))
                           .Append(mlContext.Transforms.Conversion.MapValueToKey("Label"))
                           .Append(mlContext.Ranking.Trainers.LightGbm(trainerOptions));

            var model = pipeline.Fit(trainingData);

            var predictions = model.Transform(testData);
            var metrics = mlContext.Ranking.Evaluate(predictions, evaluatorOptions,
                                                     labelColumnName: "Label", 
                                                     rowGroupColumnName: "GroupId",
                                                     scoreColumnName: "Score");

            mlContext.Model.Save(model, trainingData.Schema, $"../Checkpoints/LightGbm.zip");

            return metrics.NormalizedDiscountedCumulativeGains;
        }
    }
}

Is this expected behavior, or am I missing something ?

Thanks.

**Comment by Sumanth9Nandamuri** [comment]
The issue is caused due to missing **keyOrdinality: Microsoft.ML.Transforms.ValueToKeyMappingEstimator.KeyOrdinality.ByValue** inside **MapValueToKey("Label")**, but I think this should be the default behavior.
